lets/lets.go

```diff
@@ -0,0 +1,228 @@
+// Copyright 2022 Mineiros GmbH
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package lets
+
+import (
+	hhcl "github.com/hashicorp/hcl/v2"
+	"github.com/hashicorp/hcl/v2/hclsyntax"
+	"github.com/mineiros-io/terramate/errors"
+	"github.com/mineiros-io/terramate/hcl"
+	"github.com/mineiros-io/terramate/hcl/eval"
+	"github.com/rs/zerolog/log"
+	"github.com/zclconf/go-cty/cty"
+)
+
+// Errors returned when parsing and evaluating lets.
+const (
+	ErrLetsEval      errors.Kind = "lets eval failed"
+	ErrLetsRedefined errors.Kind = "lets redefined"
+)
+
+type (
+	// Expr is an unevaluated let expression.
+	Expr struct {
+		// Origin is the filename where this expression can be found.
+		Origin string
+
+		hclsyntax.Expression
+	}
+
+	// Exprs is the map of unevaluated let expressions visible in a
+	// directory.
+	Exprs map[string]Expr
+
+	// Value is an evaluated let.
+	Value struct {
+		Origin string
+
+		cty.Value
+	}
+
+	// L is an evaluated lets map.
+	L map[string]Value
+)
+
+// Load loads all the lets from the hcl blocks.
+func Load(letblocks hclsyntax.Blocks, ctx *eval.Context) error {
+	logger := log.With().
+		Str("action", "lets.Load()").
+		Logger()
+
+	logger.Trace().Msg("loading expressions")
+
+	exprs, err := loadExprs(letblocks)
+	if err != nil {
+		return err
+	}
+
+	return exprs.Eval(ctx)
+}
+
+// Eval evaluates all lets expressions and returns an EvalReport..
+func (letExprs Exprs) Eval(ctx *eval.Context) error {
+	logger := log.With().
+		Str("action", "Exprs.Eval()").
+		Logger()
+
+	lets := L{}
+	pendingExprsErrs := map[string]*errors.List{}
+	pendingExprs := make(Exprs)
+
+	copyexprs(pendingExprs, letExprs)
+	removeUnset(pendingExprs)
```

Comment:
Hmm do we support unset on lets ? the scope of lets is so small that it doesn't seem useful, but it makes it consistent with globals :thinking: